cranelift/isle/isle/src/trie_again.rs

```diff
@@ -0,0 +1,543 @@
+            errors.extend(self.unmatchable.drain(..).map(|err| {
+                let src = Source::new(
+                    tyenv.filenames[err.pos.file].clone(),
+                    tyenv.file_texts[err.pos.file].clone(),
+                );
+                Error::UnmatchableError {
+                    msg: format!(
+                        "rule requires binding to match both {:?} and {:?}",
+                        err.constraint_a, err.constraint_b
+                    ),
+                    src,
+                    span: Span::new_single(err.pos),
+                }
+            }))
+        }
+    }
+
+    // If a binding site is constrained and also required to be equal to another binding site, then
+    // copy the constraint and push the equality inside it.
+    fn normalize_equivalence_classes(&mut self) {
+        let mut constraints = Vec::new();
+        for binding in self.current_rule.equals.keys() {
+            if let Some(constraint) = self.current_rule.constraints.get(binding) {
+                constraints.push((*binding, *constraint));
+            }
+        }
+
+        while let Some((mut binding, constraint)) = constraints.pop() {
+            struct Pending {
+                binding: BindingId,
+                variant: sema::VariantId,
+                field: TupleIndex,
+            }
+            let mut bindings = if let Constraint::Variant(fields, _, variant) = constraint {
+                (0..fields)
+                    .map(|field| Pending {
+                        binding: self.dedup_binding(Binding::Variant {
+                            source: binding,
+                            variant,
+                            field,
+                        }),
+                        variant,
+                        field,
+                    })
+                    .collect()
+            } else {
+                Vec::new()
+            };
+
+            // Processing the first constraint in an equivalence class removes the equivalence
+            // class. If there were more constraints in that class, then we discard them here. But
+            // we left the original constraints in place above, so we've already checked that all
+            // the constraints are equal.
+            while let Some(next) = self.current_rule.equals.remove(&binding) {
+                for &mut Pending {
+                    ref mut binding,
+                    variant,
+                    field,
+                } in bindings.iter_mut()
+                {
+                    let next = self.dedup_binding(Binding::Variant {
+                        source: next,
+                        variant,
+                        field,
+                    });
+
+                    self.current_rule.equals.insert(*binding, next);
+                    if let Some(constraint) = self.current_rule.constraints.get(binding) {
+                        constraints.push((*binding, *constraint));
```

Comment:
OK so we're following an equivalence link, somehow rebuilding any Variants (ie renormalizing the `source`), and then pushing that constraint back on the workqueue?